src/hugr/views/sibling.rs

```diff
@@ -0,0 +1,572 @@
+    /// The underlying Hugr.
+    base: &'g Base,
+    /// The nodes of the induced subgraph.
+    nodes: Vec<Node>,
+}
+
+impl<'g, Base: HugrInternals> SiblingSubgraph<'g, Base> {
+    /// A sibling subgraph from a HUGR.
+    ///
+    /// The subgraph is given by the sibling graph of the root. If you wish to
+    /// create a subgraph from another root, wrap the argument `region` in a
+    /// [`super::SiblingGraph`].
+    ///
+    /// This panics if the sibling graph is empty.
+    pub fn from_sibling_graph(sibling_graph: &'g Base) -> Self
+    where
+        Base: HugrView,
+    {
+        let root = sibling_graph.root();
+        let nodes = sibling_graph.children(root).collect_vec();
+        if nodes.is_empty() {
+            panic!("Empty graph");
+        }
+        Self {
+            base: sibling_graph,
+            nodes,
+        }
+    }
+
+    /// A sibling subgraph from a [`crate::ops::OpTag::DataflowParent`]-rooted HUGR.
+    ///
+    /// The subgraph is given by the nodes between the input and output
+    /// children nodes of the parent node. If you wish to create a subgraph
+    /// from another root, wrap the `region` argument in a [`super::SiblingGraph`].
+    ///
+    /// This panics if the DFG graph is empty.
+    pub fn from_dataflow_graph(dfg_graph: &'g Base) -> Self
+    where
+        Base: HugrView,
+    {
+        let parent = dfg_graph.root();
+        let nodes = dfg_graph.children(parent).skip(2).collect_vec();
+        if nodes.is_empty() {
+            panic!("Empty DFG graph");
+        }
+        Self {
+            base: dfg_graph,
+            nodes,
+        }
+    }
+
+    /// Create a new sibling subgraph from some boundary edges.
+    ///
+    /// Any sibling subgraph can be defined using two sets of boundary edges
+    /// $B_I$ and $B_O$, the incoming and outgoing boundary edges respectively.
+    /// Intuitively, the sibling subgraph is all the edges and nodes between
+    /// an edge of $B_I$ and an edge of $B_O$.
+    ///
+    /// The `incoming` and `outgoing` arguments give $B_I$ and $B_O$ respectively.
+    /// They can be either source or target ports. We currently assume that if
+    /// the source port of an outgoing boundary edge is linked to multiple
+    /// target ports, then all edges from the same source port are outgoing
+    /// boundary edges.
+    ///
+    /// More fomally, the sibling subgraph defined by $B_I$ and $B_O$ is the
+    /// graph given by the connected components of the graph with edges
+    /// $E\B_I\B_O$ which contain at least one node that is either
+    ///  - the target of an incoming boundary edge, or
+    ///  - the source of an outgoing boundary edge.
+    ///
+    /// A subgraph is well-formed if every edge in B into the subgraph is an
+    /// incoming boundary edge and every edge in B pointing out of the subgraph
+    /// is an outgoing boundary edge.
+    ///
+    /// This function fails if the subgraph if it is not convex, if the nodes
```

Comment:
Is there an additional requirement that `incoming` and `outgoing` are disjoint?

Reply:
ok, it'd be slightly nastier than that - you can now identify a single edge by *either* endpoint...

Reply:
That's not an additional requirement: that incoming and outgoing edges are disjoint is guaranteed by convexity.

Reply:
So the concept here seems a bit ill-defined (convexity is a porperty of a subgraph, so first you have to define what nodes/edges are in the subgraph, which seems to rule out incoming+outgoing having edges in common). However, if the code here will detect that using ConvexityChecker, that's good enough for me! :)

Reply:
grammar nit tho - "if the subgraph if it is" - drop the "if it"

Reply:
```suggestion
    /// This function fails if the subgraph is not convex, if the nodes
```